RHIDP-1214: Added a procedure to configure a TLS connection with Kube… (#362)
redhat-developer/red-hat-developers-documentation-rhdh: modules/installation/proc-configuring-an-rhdh-instance-with-tls-in-kubernetes.adoc

[id="proc-configuring-an-rhdh-instance-with-tls-in-kubernetes_{context}"]
= Configuring an {product-very-short} instance with a TLS connection in Kubernetes

> You can configure an {product-very-short} instance with a Transport Layer Security (TLS) connection in your Kubernetes cluster, such as an Azure Red Hat OpenShift (ARO) cluster. You can also use a public Certificate Authority (CA)-signed certificate to secure your {product-very-short} instance.

Reviewer: Small note: my understanding is that you can **only** target a cluster with a public ca-signed certificate (such as ARO or other cloud providers or your own with proper configuration), as self-signed certificates are not supported in this case.

Author: It is a madatory thing to have a cluster configured with public Certificate Authority (CA)-signed certificate.

Modified the text:

You can configure an {product-very-short} instance with a Transport Layer Security (TLS) connection in a Kubernetes cluster, such as an Azure Red Hat OpenShift (ARO) cluster, any other cloud provider cluster, or your own cluster with proper configuration. However, You must use a public Certificate Authority (CA)-signed certificate to configure your Kubernetes cluster.

> You can configure an {product-very-short} instance with a Transport Layer Security (TLS) connection in a Kubernetes cluster, such as an Azure Red Hat OpenShift (ARO) cluster, any other cloud provider cluster, or your own cluster with proper configuration. However, You must use a public Certificate Authority (CA)-signed certificate to configure your Kubernetes cluster.

Reviewer 2: Can we say "any cluster from a supported cloud provider"?  We don't support everything

.Prerequisites

* You have set up an Azure Red Hat OpenShift (ARO) cluster with a public CA-signed certificate. For more information about obtaining CA certificates, refer to your vendor documentation.

> * You have created a namespace and setup a service account with proper read permissions on resources. For more information, see link:https://backstage.io/docs/features/kubernetes/configuration#role-based-access-control[Role-based access control].

Reviewer 3: If we're referring to it JUST so people can see the kube manifest, we should have that in as part of the RHDH docs.  Its not just verifying the content linked it, its having an assurance that its going to outlast the lifespan of our release, without change.

* You have obtained the secret and the service CA certificate associated with your service account.
* You have created some resources and added labels to them so they can be discovered by the Kubernetes plugin. For more information, see link:https://backstage.io/docs/features/kubernetes/configuration#common-backstageiokubernetes-id-label[Common `backstage.io/kubernetes-id` label].

.Procedure

. Enable the Kubernetes plugins in the `dynamic-plugins-rhdh.yaml` ConfigMap file:
+
[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
  name: dynamic-plugins-rhdh
data:
  dynamic-plugins.yaml: |
    includes:
      - dynamic-plugins.default.yaml
    plugins:
      - package: ./dynamic-plugins/dist/backstage-plugin-kubernetes-backend-dynamic
        disabled: false <1>

>       - package: ./dynamic-plugins/dist/backstage-plugin-kubernetes

Reviewer 2: The kubernetes frontend plugin is TP.  Should we be documenting TP plugins?  Maybe using the topology plugin is an alternative?

Author: Is this the plugin `./dynamic-plugins/dist/janus-idp-backstage-plugin-topology-dynamic` you are talking about? Thanks

Reviewer 2: ok, let's highlight the fact that the kubernetes frontend plugin is TP.   The topology plugin is an alternative that can be a permanent fixture since it is GA level.

          cluster: openshift
          processor:
            namespaceOverride: default
            defaultOwner: guests
          schedule:
            frequency:
              seconds: 30
            timeout:
              seconds: 5
  kubernetes:
    serviceLocatorMethod:
      type: 'multiTenant'
    clusterLocatorMethods:
      - type: 'config'
        clusters:
          - url: <target-cluster-api-server-url> <1>
            name: openshift
            authProvider: 'serviceAccount'
            skipTLSVerify: false <2>
            skipMetricsLookup: true
            dashboardUrl: <target-cluster-console-url> <3>
            dashboardApp: openshift
            serviceAccountToken: <service-account-token> <4>
            caData: <service-ca-base64-encoded> <5>
            # ...
----
<1> The base URL to the Kubernetes control plane. You can run the `kubectl cluster-info` command to get the base URL.
<2>  Set the value of this parameter to `false` to enable the verification of the TLS certificate.
<3> Optional: The link to the Kubernetes dashboard managing the ARO cluster. 
<4>  Optional: The service account token for using the `serviceAccount` authentication provider.

> <5> Set the value of the parameter to a base64-encoded certificate authority bundle in the .PEM format. 

Reviewer: Should we probably suggest to use environment variables from a secret instead of directly putting sensitive data in the configMap?

Author: Rather than suggesting, can we include an extra step to let the user know that they can create a certificate scret and use it to configure the kuberneted cluster? I have made this change, could you please have a look? Thanks

Reviewer: Is this supported? When I looked at the [backstage docs](https://github.com/backstage/backstage/blob/master/docs/features/kubernetes/configuration.md#clusterscadata-optional) my understanding was that the `caData` would require the base64-encoded certificate authority bundle in PEM format directly, not a reference to a secret. My suggestion was to simply do something like [this](https://github.com/backstage/backstage/blob/master/docs/features/kubernetes/configuration.md#configuring-kubernetes-clusters): specify `caData: ${K8S_CONFIG_CA_DATA}` where `K8S_CONFIG_CA_DATA` is just an environment variable defined in the usual rhdh-secret, like we suggest in other docs. Same for the other variables, like token and clusterName.

Author: Thank you for the clarification. 

Changed the description:
<5> Pass the CA data using a `K8S_CONFIG_CA_DATA` environment variable that you can define in your `secrets-rhdh` secret.

Reviewer: Perfect. I would use the same approach for the others `<target-cluster-api-server-url>` `<target-cluster-console-url>` `<service-account-token>` variables, or the token at least.

Author: Made this change and added a variable only for the service account token because it contains sensitive data. I think we can leave the part of defining a variable for rest of the paramters to our readers.

. Save the configuration changes.

.Verification

. Run the RHDH application to import your catalog:
+
[source,terminal]
----
kubectl -n rhdh-operator get pods -w
----

. Verify that the pod log shows no errors for your configuration.

> . Navigate to your catalog entity to view the resources that you created in a Kubernetes tab.

Author: did we check this in the ARO cluster UI? do we have any command that we can run using CLI to navigate to this catalog entity?
This is needed so that we can make the verification procedure CLI-specific, rather than performing some steps with CLI and others with web UI.

Reviewer: This is intended as checking the Entity page from the Catalog in the Developer Hub portal to ensure that the connection with the cluster is established successfully, the resources are read correctly and shown in DH, not the ARO cluster. So I would say there's no cli alternative.

[NOTE]
====

> If you face any connection error, such as certificates errors, permissions or connection issues, you can view them in a message box on the same page or in the pod's logs.

Author: do we have any specific page name to mention here for clarity??

Reviewer: Same entity page as above: instead of showing your Kubernetes/Openshift resources, it will show a message with any error occurring when fetching the data. The spefic page depends on what entity you have configured to have kubernetes/openshift resources linked to it.

.Prerequisites

* You have set up an Azure Red Hat OpenShift (ARO) cluster with a public CA-signed certificate. For more information about obtaining CA certificates, refer to your vendor documentation.
* You have created a namespace and setup a service account with proper read permissions on resources.
+
.Example: Kubernetes manifest for role-based access control
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: backstage-read-only
rules:
  - apiGroups:
      - '*'
    resources:
      - pods
      - configmaps
      - services
      - deployments
      - replicasets
      - horizontalpodautoscalers
      - ingresses
      - statefulsets
      - limitranges
      - resourcequotas
      - daemonsets
    verbs:
      - get
      - list
      - watch
#...
----

* You have obtained the secret and the service CA certificate associated with your service account.

> * You have created some resources and added annotations to them so they can be discovered by the Kubernetes plugin. You can apply these Kuberenetes annotations:

Reviewer 2: There's a typo on the second Kubernetes

** `backstage.io/kubernetes-id` to label components
** `backstage.io/kubernetes-namespace` to label namespaces

.Procedure

> . Enable the Kubernetes plugins in the `dynamic-plugins-rhdh.yaml` ConfigMap file:

Reviewer 4: How about we remove the ConfigMap since we're mentioning the file name?
```suggestion
. Enable the Kubernetes plugins in the `dynamic-plugins-rhdh.yaml` file:
```

+
[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
  name: dynamic-plugins-rhdh
data:
  dynamic-plugins.yaml: |
    includes:
      - dynamic-plugins.default.yaml
    plugins:
      - package: ./dynamic-plugins/dist/backstage-plugin-kubernetes-backend-dynamic
        disabled: false <1>
      - package: ./dynamic-plugins/dist/backstage-plugin-kubernetes
        disabled: false <2>
        # ...
----

> <1> Set the value to `false` to enable the `backstage-plugin-kubernetes-backend-dynamic` plugin

Reviewer 4: ```suggestion
<1> Set the value to `false` to enable the `backstage-plugin-kubernetes-backend-dynamic` plugin.
```

<2> Set the value to `false` to enable the `backstage-plugin-kubernetes` plugin.
+
[NOTE]
====
The `backstage-plugin-kubernetes` plugin is currently in link:https://access.redhat.com/support/offerings/techpreview[Technology Preview]. As an alternative, you can use the `./dynamic-plugins/dist/backstage-plugin-topology-dynamic` plugin, which is Generally Available (GA).
====

> . Set the kubernetes cluster details and configure the catalog sync options in the `app-config-rhdh.yaml` ConfigMap file:

Reviewer 4: Please see my previous comment about removing the ConfigMap.

            name: openshift
            authProvider: 'serviceAccount'
            skipTLSVerify: false <2>
            skipMetricsLookup: true
            dashboardUrl: <target-cluster-console-url> <3>
            dashboardApp: openshift
            serviceAccountToken: ${K8S_SERVICE_ACCOUNT_TOKEN} <4>
            caData: ${K8S_CONFIG_CA_DATA} <5>
            # ...
----
<1> The base URL to the Kubernetes control plane. You can run the `kubectl cluster-info` command to get the base URL.
<2> Set the value of this parameter to `false` to enable the verification of the TLS certificate.
<3> Optional: The link to the Kubernetes dashboard managing the ARO cluster.
<4> Optional: Pass the service account token using a `K8S_SERVICE_ACCOUNT_TOKEN` environment variable that you can define in your `secrets-rhdh` secret.
<5> Pass the CA data using a `K8S_CONFIG_CA_DATA` environment variable that you can define in your `secrets-rhdh` secret.

. Save the configuration changes.

.Verification

. Run the {product-very-short} application to import your catalog:
+
[source,terminal]
----
kubectl -n rhdh-operator get pods -w
----

. Verify that the pod log shows no errors for your configuration.

> . Navigate to the *Catalog* section and check the component page in the {product-short} portal to verify that the connection with the cluster is established and the resources that you created are present.

Reviewer 4: ```suggestion
. Go to *Catalog* and check the component page in the {product-short} instance to verify the cluster connection and the presence of your created resources.
```

[NOTE]
====

> If you face any connection error, such as certificate errors, permissions or connection issues, you can view them in a message box on the same component page or in the pod's logs.

Reviewer 4: ```suggestion
If you encounter connection errors, such as certificate issues or permissions, check the message box in the component page or in the pod's logs.
```
